File: src/image_generator.py
import base64
import hashlib
import io
import time
from pathlib import Path
import requests
from urllib.parse import quote
from PIL import Image
from google import genai
from google.genai import types
from .config import CACHE_DIR, GEMINI_API_KEY, VIDEO_WIDTH, VIDEO_HEIGHT

GEMINI_IMAGE_MODEL = "gemini-3.1-flash-image-preview"
POLLINATIONS_URL = "https://image.pollinations.ai/prompt/{prompt}"


def _cache_path(prompt: str, seed: int, provider: str) -> Path:
    key = hashlib.sha256(f"{provider}|{prompt}|{seed}".encode()).hexdigest()[:20]
    return CACHE_DIR / f"img_{provider}_{key}.jpg"


def _gemini_generate(prompt: str, out_path: Path, aspect_ratio: str = "16:9") -> Path:
    if not GEMINI_API_KEY:
        raise RuntimeError("GEMINI_API_KEY não configurada")

    client = genai.Client(api_key=GEMINI_API_KEY)

    full_prompt = (
        f"Generate a single cinematic, photorealistic image in {aspect_ratio} aspect ratio. "
        f"High production quality, professional cinematography, dramatic lighting, sharp focus, "
        f"detailed textures. No text, no captions, no watermarks in the image.\n\n"
        f"Scene: {prompt}"
    )

    response = client.models.generate_content(
        model=GEMINI_IMAGE_MODEL,
        contents=full_prompt,
        config=types.GenerateContentConfig(
            response_modalities=["IMAGE"],
            image_config=types.ImageConfig(aspect_ratio=aspect_ratio),
        ),
    )

    for part in response.candidates[0].content.parts:
        if getattr(part, "inline_data", None) and part.inline_data.data:
            data = part.inline_data.data
            if isinstance(data, str):
                data = base64.b64decode(data)
            img = Image.open(io.BytesIO(data)).convert("RGB")
            img.save(out_path, "JPEG", quality=92)
            return out_path

    raise RuntimeError("Gemini não retornou imagem")


def _pollinations_generate(
    prompt: str, seed: int, out_path: Path,
    width: int = VIDEO_WIDTH, height: int = VIDEO_HEIGHT,
) -> Path:
    full_prompt = f"{prompt}, cinematic, photorealistic, 8k, high detail"
    url = POLLINATIONS_URL.format(prompt=quote(full_prompt))
    params = {
        "width": width, "height": height, "seed": seed,
        "nologo": "true", "enhance": "true", "model": "flux",
    }
    r = requests.get(url, params=params, timeout=120)
    r.raise_for_status()
    if len(r.content) < 1000:
        raise RuntimeError("Imagem retornada muito pequena")
    out_path.write_bytes(r.content)
    return out_path


def generate_image(
    prompt: str,
    seed: int = 42,
    aspect_ratio: str = "16:9",
    provider: str = "gemini",
    retries: int = 3,
) -> Path:
    cache = _cache_path(prompt, seed, provider)
    if cache.exists() and cache.stat().st_size > 1000:
        return cache

    last_err = None
    for attempt in range(retries):
        try:
            if provider == "gemini":
                return _gemini_generate(prompt, cache, aspect_ratio=aspect_ratio)
            else:
                w, h = (VIDEO_WIDTH, VIDEO_HEIGHT) if aspect_ratio == "16:9" else (1080, 1920)
                return _pollinations_generate(prompt, seed, cache, width=w, height=h)
        except Exception as e:
            last_err = e
            time.sleep(2 ** attempt)

    if provider == "gemini":
        try:
            print(f"[image_generator] Gemini falhou, usando Pollinations: {last_err}")
            w, h = (VIDEO_WIDTH, VIDEO_HEIGHT) if aspect_ratio == "16:9" else (1080, 1920)
            fallback_cache = _cache_path(prompt, seed, "pollinations")
            if fallback_cache.exists() and fallback_cache.stat().st_size > 1000:
                return fallback_cache
            return _pollinations_generate(prompt, seed, fallback_cache, width=w, height=h)
        except Exception as e:
            last_err = e

    raise RuntimeError(f"Falha ao gerar imagem após {retries} tentativas: {last_err}")

File: src/lib/pipeline/image-generator.ts
import { createHash } from "node:crypto";
import { existsSync, statSync } from "node:fs";
import { mkdir, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { getGenAI, MODELS } from "@/lib/gemini/client";
import type { ImageProvider, VideoFormat } from "@/lib/types/jobs";
import { getCacheDir } from "@/lib/utils/paths";
import { sleep } from "@/lib/utils/time";

const POLLINATIONS_BASE =
  process.env.POLLINATIONS_BASE_URL ?? "https://image.pollinations.ai/prompt";

function cachePath(prompt: string, seed: number, provider: ImageProvider): string {
  const key = createHash("sha256")
    .update(`${provider}|${prompt}|${seed}`)
    .digest("hex")
    .slice(0, 20);
  return join(getCacheDir(), `img_${provider}_${key}.jpg`);
}

function isCacheValid(path: string): boolean {
  if (!existsSync(path)) return false;
  try {
    return statSync(path).size > 1000;
  } catch {
    return false;
  }
}

async function geminiGenerate(
  prompt: string,
  outPath: string,
  aspectRatio: VideoFormat,
): Promise<string> {
  const ai = getGenAI();

  const fullPrompt =
    `Generate a single cinematic, photorealistic image in ${aspectRatio} aspect ratio. ` +
    `High production quality, professional cinematography, dramatic lighting, sharp focus, ` +
    `detailed textures. No text, no captions, no watermarks in the image.\n\n` +
    `Scene: ${prompt}`;

  const response = await ai.models.generateContent({
    model: MODELS.image,
    contents: fullPrompt,
    config: {
      responseModalities: ["IMAGE"],
      imageConfig: { aspectRatio },
    },
  });

  const parts = response.candidates?.[0]?.content?.parts ?? [];
  for (const part of parts) {
    const inline = part.inlineData;
    if (inline?.data) {
      const buffer = Buffer.from(inline.data as string, "base64");
      await writeFile(outPath, buffer);
      return outPath;
    }
  }
  throw new Error("Gemini não retornou imagem");
}

async function pollinationsGenerate(
  prompt: string,
  seed: number,
  outPath: string,
  width: number,
  height: number,
): Promise<string> {
  const fullPrompt = `${prompt}, cinematic, photorealistic, 8k, high detail`;
  const url = new URL(`${POLLINATIONS_BASE}/${encodeURIComponent(fullPrompt)}`);
  url.searchParams.set("width", String(width));
  url.searchParams.set("height", String(height));
  url.searchParams.set("seed", String(seed));
  url.searchParams.set("nologo", "true");
  url.searchParams.set("enhance", "true");
  url.searchParams.set("model", "flux");

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 120_000);
  try {
    const r = await fetch(url, { signal: controller.signal });
    if (!r.ok) throw new Error(`Pollinations HTTP ${r.status}`);
    const buf = Buffer.from(await r.arrayBuffer());
    if (buf.byteLength < 1000) {
      throw new Error("Imagem retornada muito pequena");
    }
    await writeFile(outPath, buf);
    return outPath;
  } finally {
    clearTimeout(timeout);
  }
}

export interface GenerateImageOptions {
  prompt: string;
  seed?: number;
  aspectRatio?: VideoFormat;
  provider?: ImageProvider;
  retries?: number;
}

export async function generateImage({
  prompt,
  seed = 42,
  aspectRatio = "16:9",
  provider = "gemini",
  retries = 3,
}: GenerateImageOptions): Promise<string> {
  await mkdir(getCacheDir(), { recursive: true });

  const cache = cachePath(prompt, seed, provider);
  if (isCacheValid(cache)) return cache;

  const [width, height] = aspectRatio === "16:9" ? [1920, 1080] : [1080, 1920];

  let lastErr: Error | null = null;
  for (let attempt = 0; attempt < retries; attempt++) {
    try {
      if (provider === "gemini") {
        return await geminiGenerate(prompt, cache, aspectRatio);
      }
      return await pollinationsGenerate(prompt, seed, cache, width, height);
    } catch (err) {
      lastErr = err as Error;
      await sleep(1000 * 2 ** attempt);
    }
  }

  if (provider === "gemini") {
    // fallback automático para Pollinations
    try {
      const fallbackCache = cachePath(prompt, seed, "pollinations");
      if (isCacheValid(fallbackCache)) return fallbackCache;
      return await pollinationsGenerate(prompt, seed, fallbackCache, width, height);
    } catch (err) {
      lastErr = err as Error;
    }
  }

  throw new Error(
    `Falha ao gerar imagem após ${retries} tentativas: ${lastErr?.message ?? "desconhecido"}`,
  );
}
